### ISSUE TITLE
Too many parameters causes segfault on launch

### ISSUE THREAD

**Issue by ben-willmore** [libretro/mame#513]
I am using mame_libretro.so on an aarch64 anbernic handheld running knulli linux. When using the BBC Micro emulator, there is a segfault on launch. This can be fixed by removing certain parameters from the command line.

This worked fine on 0.265 and is broken on 0.268 and later.

I launch retroarch with:
`/usr/bin/retroarch -L /usr/lib/libretro/mame_libretro.so --verbose ./bad.cmd`
--> segmentation fault, log attached

File `bad.cmd` contains:
```
bbcb -flop1 "/userdata/roms/bbc/3D Bomb Alley (Software Invasion) (1983).ssd" -r
ompath "/userdata/roms/bbc;/userdata/bios/" -cfg_directory /userdata/saves/mame/
cfg/bbcb -inipath /userdata/saves/mame/mame/ini -artpath /var/run/mame_artwork/;
/usr/bin/mame/artwork/;/userdata/bios/lr-mame/artwork;/userdata/bios/mame/artwor
k;/userdata/decorations -artwork_crop -pluginspath /usr/bin/mame/plugins/;/userd
ata/saves/mame/plugins -homepath /userdata/saves/mame/plugins -samplepath /userd
ata/bios/mame/samples
```

If I remove one pair of parameters from the cmd file, the problem is fixed:

File `good.cmd` contains:
```
bbcb -flop1 "/userdata/roms/bbc/3D Bomb Alley (Software Invasion) (1983).ssd" -r
ompath "/userdata/roms/bbc;/userdata/bios/" -cfg_directory /userdata/saves/mame/
cfg/bbcb -inipath /userdata/saves/mame/mame/ini -artpath /var/run/mame_artwork/;
/usr/bin/mame/artwork/;/userdata/bios/lr-mame/artwork;/userdata/bios/mame/artwor
k;/userdata/decorations -artwork_crop -pluginspath /usr/bin/mame/plugins/;/userd
ata/saves/mame/plugins -homepath /userdata/saves/mame/plugins
```

`/usr/bin/retroarch -L /usr/lib/libretro/mame_libretro.so --verbose ./good.cmd`
--> (one less pair of parameters) --> , emulator launches correctly

I can also fix the problem by slightly adjusting the filename parameter, by removing the initial "./". Different forms of the path either do or don't cause the crash:

`/usr/bin/retroarch -L /usr/lib/libretro/mame_libretro.so --verbose bad.cmd`
--> (no ./ before the filename) --> success, emulator launches correctly.

[retroarch.log](https://github.com/user-attachments/files/18183912/retroarch.log)

**Comment by sonninnos** [comment]
The bad string is 513 bytes, so there must be a 512 sized buffer somewhere..

Edit: try this if you can build it:
```diff
diff --git a/src/osd/libretro/libretro-internal/retro_init.cpp b/src/osd/libretro/libretro-internal/retro_init.cpp
index 8eb046a9ecc..ab6dc81ac4c 100755
--- a/src/osd/libretro/libretro-internal/retro_init.cpp
+++ b/src/osd/libretro/libretro-internal/retro_init.cpp
@@ -898,7 +898,7 @@ static int execute_game_cmd(char *path)

 #include <fstream>
 #include <string>
-static char CMDFILE[512];
+static char CMDFILE[1024];

 int loadcmdfile(char *argv)
 {
```

The crash is because it is using `sprintf` to read the string, and it has no size limit. Should be doing `snprintf` instead, so that it gracefully cuts the string.

**Comment by ben-willmore** [comment]
You're right! However, just increasing CMDFILE doesn't fix the test case. The following seems to:
```
diff --git a/src/osd/libretro/libretro-internal/retro_init.cpp b/src/osd/libretro/libretro-internal/retro_init.cpp
index 8eb046a9ecc..8376e56f4cf 100755
--- a/src/osd/libretro/libretro-internal/retro_init.cpp
+++ b/src/osd/libretro/libretro-internal/retro_init.cpp
@@ -673,7 +673,7 @@ static int execute_game(char *path)
 static void parse_cmdline(const char *argv)
 {
    int c,c2;
-   static char buffer[512*4];
+   static char buffer[1024*4];
    enum states
    {
       DULL,
@@ -898,7 +898,7 @@ static int execute_game_cmd(char *path)
 
 #include <fstream>
 #include <string>
-static char CMDFILE[512];
+static char CMDFILE[1024];
 
 int loadcmdfile(char *argv)
 {
@@ -953,14 +953,14 @@ int mmain2(int argc, const char *argv)
    }
    else
    {
-      char argv_trimmed[512];
+      char argv_trimmed[1024];
 
       argv_trimmed[0] = '\0';
 
       if (argv[0])
       {
-         char argv_first[512];
-         char a[512];
+         char argv_first[1024];
+         char a[1024];
          const char *first = NULL;
 
          strcpy(argv_first, argv);
```

**Comment by sonninnos** [comment]
Does not make any sense why those others are needed for the test case since the first is already big enough, and those in `mmain2` are only used when starting with an actual command (as in the cmd-file contents directly) and not a cmd-file..

**Comment by ben-willmore** [comment]
I will double-check that all of those are needed, but are these errors relevant? It seems strange that it is attempting to interpret "./bad.cmd"  as a system name:

```
[libretro ERROR] Error parsing system name: "./bad.cmd"
[libretro ERROR] Error parsing parent path: "./bad.cmd"
```

**Comment by ben-willmore** [comment]
OK, I checked each one and the mmain2 changes are **not** needed  but the other two are:

```
diff --git a/src/osd/libretro/libretro-internal/retro_init.cpp b/src/osd/libretro/libretro-internal/retro_init.cpp
index 8eb046a9ecc..c1463703b76 100755
--- a/src/osd/libretro/libretro-internal/retro_init.cpp
+++ b/src/osd/libretro/libretro-internal/retro_init.cpp
@@ -673,7 +673,7 @@ static int execute_game(char *path)
 static void parse_cmdline(const char *argv)
 {
    int c,c2;
-   static char buffer[512*4];
+   static char buffer[1024*4];
    enum states
    {
       DULL,
@@ -898,7 +898,7 @@ static int execute_game_cmd(char *path)
 
 #include <fstream>
 #include <string>
-static char CMDFILE[512];
+static char CMDFILE[1024];
 
 int loadcmdfile(char *argv)
 {
```
Any value less than (1024*4)-10 for buffer causes a segfault on the test case.

**Comment by ben-willmore** [comment]
OK, I have now worked this out. The size of the cmd file is not critical, I can make the file much short (260 chars) with the same number of parameters, and it will still fail. However, the number of parameters is critical.

I printed out the arguments at line 1006 of `retro_init.cpp`:
```
         xargv_cmd[i] = (char*)(XARGV[i]);
         log_cb(RETRO_LOG_INFO, "%d:  %s\n", i, XARGV[i]);
```

Here are the results:

Failure case: `/usr/bin/retroarch -L ./mame_libretro.so --verbose ./bad.cmd`
```
[libretro INFO] 0:  mame
...
[libretro INFO] 65:  /userdata/bios/mame/samples
```

Success case #1: `/usr/bin/retroarch -L ./mame_libretro.so --verbose bad.cmd`
```
[libretro INFO] 0:  mame
...
[libretro INFO] 63:  /userdata/bios/mame/samples
```

Success case #2: `/usr/bin/retroarch -L ./mame_libretro.so --verbose ./good.cmd`
```
[libretro INFO] 0:  mame
...
[libretro INFO] 63:  /userdata/bios/mame/samples
```

Note that the success cases have 64 arguments and the failure case has 66 arguments. The size of XARGV and xargv_cmd is 64. Changing these to 96 corrects the segfault.

[retroarch.patch](https://github.com/user-attachments/files/18189574/retroarch.patch)
